File: app/electron/core/settings-manager.js
import fs from 'node:fs';
import path from 'node:path';
import os from 'node:os';
import { fileURLToPath } from 'node:url';

const log = (...args) => {
  console.log(`[${new Date().toISOString()}]`, ...args);
};

const DEFAULT_SETTINGS = {
  pauseBeforeStop: 500,
  connection: {
    type: 'usb',
    ip: '192.168.5.1',
    port: 23,
    serverPort: 8090,
    usbPort: '',
    baudRate: 115200
  },
  theme: 'dark',
  workspace: 'G54',
  defaultGcodeView: 'top',
  autoFit : false,
  accentColor: '#1abc9c',
  gradientColor: '#34d399',
  autoClearConsole: true,
  debugLogging: false,
  unitsPreference: 'metric',
  homeLocation: 'back-left',
  keyboardBindings: {
    ArrowUp: 'JogYPlus',
    ArrowDown: 'JogYMinus',
    ArrowLeft: 'JogXMinus',
    ArrowRight: 'JogXPlus',
    PageUp: 'JogZPlus',
    PageDown: 'JogZMinus',
    Escape: 'JogCancel'
  },
  keyboard: {
    shortcutsEnabled: true,
    step: 1,
    xyFeedRate: 3000,
    zFeedRate: 1500
  },
  features: {
    keyboardShortcuts: true
  },
  consoleBufferSize: 1000,
  lastLoadedFile: null,
  enableStateDeltaBroadcast: true,
  probe: {
    type: 'autozero-touch',
    probingAxis: 'Z',
    selectedCorner: 'BottomLeft',
    typeInitialized: false,
    requireConnectionTest: false,
    bitDiameters: [2, 3.175, 6.35, 9.525, 12],
    '3d-probe': {
      ballPointDiameter: 6,
      zPlunge: 3,
      zOffset: 0,
      rapidMovement: 2000,
      xDimension: 100,
      yDimension: 100,
      probeZFirst: false
    },
    'standard-block': {
      zThickness: 25,
      xyThickness: 50,
      zProbeDistance: 3
    }
  },
  tool: {
    count: 0,
    source: null,
    tls: false,
    manual: false
  },
  plugins: {
    allowPriorityReordering: false
  },
  useDoorAsPause : true
};

function getUserDataDir() {
  const platform = os.platform();
  const appName = 'ncSender';

  switch (platform) {
    case 'win32':
      return path.join(os.homedir(), 'AppData', 'Roaming', appName);
    case 'darwin':
      return path.join(os.homedir(), 'Library', 'Application Support', appName);
    case 'linux':
      return path.join(os.homedir(), '.config', appName);
    default:
      return path.join(os.homedir(), `.${appName}`);
  }
}

const SETTINGS_PATH = path.join(getUserDataDir(), 'settings.json');

function ensureSettingsFile() {
  if (!fs.existsSync(SETTINGS_PATH)) {
    try {
      // Ensure directory exists
      const settingsDir = path.dirname(SETTINGS_PATH);
      fs.mkdirSync(settingsDir, { recursive: true });
      fs.writeFileSync(SETTINGS_PATH, JSON.stringify(DEFAULT_SETTINGS, null, 2), 'utf8');
    } catch (error) {
      log('Failed to create default settings file:', error);
    }
  }
}

export function readSettings() {
  // Check if settings file exists, but don't create it
  if (!fs.existsSync(SETTINGS_PATH)) {
    return null; // Indicate that no settings file exists
  }

  try {
    const raw = fs.readFileSync(SETTINGS_PATH, 'utf8');
    if (!raw) {
      return { ...DEFAULT_SETTINGS };
    }
    const parsed = JSON.parse(raw);
    const parsedObject = parsed && typeof parsed === 'object' ? parsed : {};
    const merged = { ...DEFAULT_SETTINGS, ...parsedObject };

    const parsedProbe = parsedObject.probe && typeof parsedObject.probe === 'object'
      ? parsedObject.probe
      : {};

    const mergedProbe = {
      ...DEFAULT_SETTINGS.probe,
      ...parsedProbe,
      '3d-probe': {
        ...DEFAULT_SETTINGS.probe['3d-probe'],
        ...(parsedProbe['3d-probe'] || {})
      },
      'standard-block': {
        ...DEFAULT_SETTINGS.probe['standard-block'],
        ...(parsedProbe['standard-block'] || {})
      }
    };

    if (mergedProbe.typeInitialized !== true && mergedProbe.type === '3d-probe') {
      mergedProbe.type = 'autozero-touch';
    }

    merged.probe = mergedProbe;

    const parsedKeyboard = parsedObject.keyboard && typeof parsedObject.keyboard === 'object'
      ? parsedObject.keyboard
      : {};

    merged.keyboard = {
      ...DEFAULT_SETTINGS.keyboard,
      ...parsedKeyboard
    };

    const parsedFeatures = parsedObject.features && typeof parsedObject.features === 'object'
      ? parsedObject.features
      : {};

    merged.features = {
      ...DEFAULT_SETTINGS.features,
      ...parsedFeatures
    };

    // Migrate old numberOfTools to new tool.count structure
    const parsedTool = parsedObject.tool && typeof parsedObject.tool === 'object'
      ? parsedObject.tool
      : {};

    merged.tool = {
      ...DEFAULT_SETTINGS.tool,
      ...parsedTool
    };

    // Migration: convert old numberOfTools to tool.count
    if (parsedObject.numberOfTools !== undefined && parsedObject.tool === undefined) {
      merged.tool.count = parsedObject.numberOfTools;
    }

    const parsedConnection = parsedObject.connection;
    const hasCustomConnection = parsedConnection && typeof parsedConnection === 'object';
    const normalizedConnection = {
      ...DEFAULT_SETTINGS.connection,
      ...(hasCustomConnection ? parsedConnection : {})
    };

    // Support legacy property name
    if (
      hasCustomConnection &&
      typeof parsedConnection.connectionType === 'string' &&
      !parsedConnection.type
    ) {
      normalizedConnection.type = parsedConnection.connectionType.toLowerCase();
    }

    // If user provided a connection object but omitted type entirely, treat as unconfigured
    const typeProvided = hasCustomConnection &&
      (Object.prototype.hasOwnProperty.call(parsedConnection, 'type') ||
       Object.prototype.hasOwnProperty.call(parsedConnection, 'connectionType'));

    if (!typeProvided) {
      delete normalizedConnection.type;
    } else if (typeof normalizedConnection.type === 'string') {
      normalizedConnection.type = normalizedConnection.type.toLowerCase();
    }

    delete normalizedConnection.connectionType;

    merged.connection = normalizedConnection;
    return merged;
  } catch (error) {
    log('Failed to load settings. Using defaults.', error);
    return { ...DEFAULT_SETTINGS };
  }
}

export function getSetting(key, fallback) {
  const settings = readSettings();

  // If no settings file exists, return undefined (no fallbacks)
  if (settings === null) {
    return undefined;
  }

  if (Object.prototype.hasOwnProperty.call(settings, key)) {
    const value = settings[key];
    if (value !== undefined) {
      return value;
    }
  }
  if (fallback !== undefined) {
    return fallback;
  }
  if (Object.prototype.hasOwnProperty.call(DEFAULT_SETTINGS, key)) {
    return DEFAULT_SETTINGS[key];
  }
  return undefined;
}

export function saveSettings(newSettings) {
  ensureSettingsFile();

  try {
    const currentSettings = readSettings();
    const updatedSettings = { ...currentSettings, ...newSettings };

    if (updatedSettings.connection && typeof updatedSettings.connection === 'object') {
      delete updatedSettings.ip;
      delete updatedSettings.port;
      delete updatedSettings.serverPort;
      delete updatedSettings.usbPort;
      delete updatedSettings.connectionType;
      delete updatedSettings.connection.connectionType;
    }
    delete updatedSettings.baudRate;

    fs.writeFileSync(SETTINGS_PATH, JSON.stringify(updatedSettings, null, 2), 'utf8');
    return updatedSettings;
  } catch (error) {
    log('Failed to save settings:', error);
    throw error;
  }
}

export function saveConnectionSettings(connectionSettings) {
  const parsedBaudRate = parseInt(connectionSettings.baudRate, 10);
  const parsedPort = parseInt(connectionSettings.port, 10);
  const parsedServerPort = parseInt(connectionSettings.serverPort, 10);

  const settingsToSave = {
    connection: {
      type: connectionSettings.type?.toLowerCase() || DEFAULT_SETTINGS.connection.type,
      ip: connectionSettings.ipAddress || DEFAULT_SETTINGS.connection.ip,
      port: Number.isFinite(parsedPort) ? parsedPort : DEFAULT_SETTINGS.connection.port,
      serverPort: Number.isFinite(parsedServerPort) ? parsedServerPort : DEFAULT_SETTINGS.connection.serverPort,
      usbPort: connectionSettings.usbPort || DEFAULT_SETTINGS.connection.usbPort,
      baudRate: Number.isFinite(parsedBaudRate) ? parsedBaudRate : DEFAULT_SETTINGS.connection.baudRate
    }
  };

  return saveSettings(settingsToSave);
}

export function removeSetting(key) {
  ensureSettingsFile();

  try {
    const currentSettings = readSettings();
    if (currentSettings && Object.prototype.hasOwnProperty.call(currentSettings, key)) {
      const { [key]: _, ...updatedSettings } = currentSettings;
      fs.writeFileSync(SETTINGS_PATH, JSON.stringify(updatedSettings, null, 2), 'utf8');
      return updatedSettings;
    }
    return currentSettings;
  } catch (error) {
    log(`Failed to remove setting '${key}':`, error);
    throw error;
  }
}

export { SETTINGS_PATH, DEFAULT_SETTINGS };

File: src/NcSender.Server/Configuration/SettingsManager.cs
using System.Text.Json;
using System.Text.Json.Nodes;
using NcSender.Core.Interfaces;
using NcSender.Server.Infrastructure;

namespace NcSender.Server.Configuration;

public class SettingsManager : ISettingsManager
{
    private readonly string _settingsPath;
    private readonly object _lock = new();
    private readonly SemaphoreSlim _writeLock = new(1, 1);
    private JsonObject _settings;

    public SettingsManager() : this(PathUtils.GetSettingsPath()) { }

    public SettingsManager(string settingsPath)
    {
        _settingsPath = settingsPath;
        _settings = LoadFromDisk();
    }

    public JsonNode? GetSetting(string key, JsonNode? fallback = null)
    {
        lock (_lock)
        {
            var value = GetNestedValue(_settings, key);
            return value ?? fallback;
        }
    }

    public T? GetSetting<T>(string key, T? fallback = default)
    {
        var node = GetSetting(key);
        if (node is null)
            return fallback;

        try
        {
            // Use GetValue<T> for primitives — works in AOT without reflection.
            // Deserialize<T> requires the System.Text.Json serializer which needs
            // source-gen metadata for each type in Native AOT builds.
            return node.GetValue<T>();
        }
        catch
        {
            return fallback;
        }
    }

    public async Task SaveSettings(JsonObject newSettings)
    {
        lock (_lock)
        {
            DeepMerge(_settings, newSettings);
        }

        await PersistToDisk();
    }

    public JsonObject ReadAll()
    {
        lock (_lock)
        {
            // Return a deep copy of the in-memory settings to prevent external mutation.
            // The in-memory state is always current via SaveSettings/DeepMerge.
            // Re-reading from disk here caused race conditions with concurrent
            // PersistToDisk writes (non-atomic File.WriteAllTextAsync could be read
            // mid-truncation, falling back to CreateDefaults and losing connection
            // settings, which triggered spurious "setup-required" status).
            return _settings.DeepClone().AsObject();
        }
    }

    public async Task RemoveSetting(string key)
    {
        lock (_lock)
        {
            RemoveNestedValue(_settings, key);
        }

        await PersistToDisk();
    }

    private JsonObject LoadFromDisk()
    {
        if (!File.Exists(_settingsPath))
            return CreateDefaults();

        try
        {
            var json = File.ReadAllText(_settingsPath);
            var parsed = JsonNode.Parse(json)?.AsObject();
            if (parsed is not null)
                return parsed;
        }
        catch
        {
            // Corrupted file — start fresh with defaults
        }

        return CreateDefaults();
    }

    private async Task PersistToDisk()
    {
        string json;
        lock (_lock)
        {
            // Use WriteTo(Utf8JsonWriter) instead of ToJsonString(JsonSerializerOptions)
            // to avoid relying on JsonSerializer which can fail in AOT when
            // JsonSerializerOptions lacks a TypeInfoResolver.
            using var ms = new System.IO.MemoryStream();
            using (var writer = new System.Text.Json.Utf8JsonWriter(ms, new System.Text.Json.JsonWriterOptions { Indented = true }))
            {
                _settings.WriteTo(writer);
            }
            json = System.Text.Encoding.UTF8.GetString(ms.ToArray());
        }

        // Serialize disk writes so concurrent PATCHes don't race on the temp file.
        await _writeLock.WaitAsync();
        try
        {
            var dir = Path.GetDirectoryName(_settingsPath)!;
            Directory.CreateDirectory(dir);

            // Atomic write: write to a temp file then rename, so a crash or concurrent
            // write never leaves a truncated/corrupt settings.json on disk.
            var tmpPath = _settingsPath + ".tmp";
            await File.WriteAllTextAsync(tmpPath, json);
            File.Move(tmpPath, _settingsPath, overwrite: true);
        }
        finally
        {
            _writeLock.Release();
        }
    }

    private static JsonNode? GetNestedValue(JsonObject obj, string path)
    {
        var parts = path.Split('.');
        JsonNode? current = obj;

        foreach (var part in parts)
        {
            if (current is not JsonObject currentObj || !currentObj.ContainsKey(part))
                return null;

            current = currentObj[part];
        }

        return current;
    }

    private static void RemoveNestedValue(JsonObject obj, string path)
    {
        var parts = path.Split('.');
        JsonObject current = obj;

        for (int i = 0; i < parts.Length - 1; i++)
        {
            if (current[parts[i]] is not JsonObject next)
                return;
            current = next;
        }

        current.Remove(parts[^1]);
    }

    /// <summary>
    /// Recursively merge source into target. Source values overwrite target values.
    /// Nested objects are merged recursively.
    /// </summary>
    private static void DeepMerge(JsonObject target, JsonObject source)
    {
        foreach (var (key, sourceValue) in source)
        {
            if (sourceValue is JsonObject sourceObj
                && target[key] is JsonObject targetObj)
            {
                DeepMerge(targetObj, sourceObj);
            }
            else
            {
                target[key] = sourceValue?.DeepClone();
            }
        }
    }

    private static JsonObject CreateDefaults()
    {
        var json = """
        {
            "pauseBeforeStop": 500,
            "pollingInterval": 100,
            "connection": {
                "type": "usb",
                "ip": "192.168.5.1",
                "port": 23,
                "protocol": "telnet",
                "serverPort": 8090,
                "usbPort": "",
                "baudRate": 115200
            },
            "theme": "dark",
            "workspace": "G54",
            "defaultGcodeView": "top",
            "autoFit": false,
            "accentColor": "#1abc9c",
            "gradientColor": "#34d399",
            "autoClearConsole": true,
            "debugLogging": false,
            "unitsPreference": "metric",
            "homeLocation": "back-left",
            "keyboardBindings": {
                "ArrowUp": "JogYPlus",
                "ArrowDown": "JogYMinus",
                "ArrowLeft": "JogXMinus",
                "ArrowRight": "JogXPlus",
                "PageUp": "JogZPlus",
                "PageDown": "JogZMinus",
                "Escape": "JogCancel"
            },
            "keyboard": {
                "shortcutsEnabled": true,
                "step": 1,
                "xyFeedRate": 3000,
                "zFeedRate": 1500
            },
            "features": {
                "keyboardShortcuts": true
            },
            "consoleBufferSize": 1000,
            "lastLoadedFile": null,
            "enableStateDeltaBroadcast": true,
            "probe": {
                "type": "autozero-touch",
                "probingAxis": "Z",
                "selectedCorner": "BottomLeft",
                "typeInitialized": false,
                "requireConnectionTest": false,
                "bitDiameters": [2, 3.175, 6.35, 9.525, 12]
            },
            "tool": {
                "count": 0,
                "source": null,
                "tls": false,
                "manual": false
            },
            "plugins": {
                "allowPriorityReordering": false
            },
            "useDoorAsPause": true,
            "remoteControl": {
                "enabled": false
            },
            "showPendant": false,
            "pendant": {
                "serialPort": "auto",
                "baudRate": 115200,
                "autoConnect": true
            }
        }
        """;

        return JsonNode.Parse(json)!.AsObject();
    }
}
